### PR TITLE
Work around poor codegen for cttz8 with defined results for zero

### PR DESCRIPTION
**Pull request by dotdash**
Currently, LLVM lowers a cttz8 on x86_64 to these instructions:

``` asm
    movzbl      %dil, %eax
    bsfl        %eax, %eax
    movl        $32, %ecx
    cmovnel     %eax, %ecx
    cmpl        $32, %ecx
    movl        $8, %eax
    cmovnel     %ecx, %eax
```

To improve the codegen, we can zero extend the 8 bit integer, then set
bit 8 and perform a cttz operation on the extended value. That way
there's no conditional operation involved at all.

This was discovered by  this benchmark: https://github.com/Kimundi/long_strings_without_repeats

Timings on my box with the current nightly:

```
running 4 tests
test bench_cpp_naive_big     ... bench:   5479222 ns/iter (+/- 254222)
test bench_noop_big          ... bench:    571405 ns/iter (+/- 111950)
test bench_rust_naive_big    ... bench:   7798102 ns/iter (+/- 148841)
test bench_rust_unsafe_big   ... bench:   6606488 ns/iter (+/- 67529)
```

Timings with the patch applied:

```
running 4 tests
test bench_cpp_naive_big     ... bench:   5470944 ns/iter (+/- 7109)
test bench_noop_big          ... bench:    568944 ns/iter (+/- 6895)
test bench_rust_naive_big    ... bench:   6795901 ns/iter (+/- 43806)
test bench_rust_unsafe_big   ... bench:   5584879 ns/iter (+/- 5291)
```

### CODE REVIEW

**Comment by dotdash** [comment]
Hmm, just noticed that the flag must be a constant, so a call like `cttz(val, some_var)` crashes the compiler. :-(

**Comment by dotdash** [comment]
For reference: http://reviews.llvm.org/D9284

**Comment by dotdash** [comment]
Updated the patch to do the same thing that the updated LLVM would do to avoid the problems with the flag possibly crashing the compiler.

**Comment by Kimundi** [comment]
Hm, I'm confused. All the patch does right now is to replace the cttz8 call with a cttz16 call, which does not seem to improve the runtime for the benchmarks at all on my end.

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -745,7 +745,18 @@ macro_rules! uint_impl {
         #[stable(feature = "rust1", since = "1.0.0")]
         #[inline]
         pub fn trailing_zeros(self) -> u32 {
-            unsafe { $cttz(self as $ActualT) as u32 }
+            // As of LLVM 3.6 the codegen for the zero-safe cttz8 intrinsic emits two conditional
+            // moves on x86_64. By promoting the value to u32 and setting bit 8, we get better code
+            // without any conditional operations.
+            // FIXME: There's a LLVM patch pending, remove this workaround once LLVM generates
```
Comment (steveklabnik) [inline src/libcore/num/mod.rs]: maybe link to the patch

**Comment by alexcrichton** [comment]
@Kimundi could you post how you're running the benchmark? With this diff against the repo @dotdash linked to:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 3f39057..b5787de 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -23,7 +23,7 @@ fn log2_ceil(n: u8) -> u8 {
 fn lsb_differ_index(a: u8, b: u8) -> u8 {
     // xor => bits with differences are 1
     // trailing_zero => position of first difference
-    (a ^ b).trailing_zeros() as u8
+    ((a ^ b) as u16 | 0x100).trailing_zeros() as u8
 }
 
 #[test]
```

I get these numbers before:

```
```

**Comment by alexcrichton** [comment]
Gah comment button! Anyway, I get these numbers before:

```
test bench_rust_naive_big    ... bench:   7438584 ns/iter (+/- 216830)              
test bench_rust_naive_short  ... bench:        82 ns/iter (+/- 3)                   
test bench_rust_unsafe_big   ... bench:   6602519 ns/iter (+/- 362681)              
test bench_rust_unsafe_short ... bench:        74 ns/iter (+/- 3)                   
```

and these numbers after

```
test bench_rust_naive_big    ... bench:   6349330 ns/iter (+/- 2569100)         
test bench_rust_naive_short  ... bench:       120 ns/iter (+/- 8)               
test bench_rust_unsafe_big   ... bench:   5299008 ns/iter (+/- 173938)          
test bench_rust_unsafe_short ... bench:        60 ns/iter (+/- 6)               
```

Which looks like a good win to me!

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -745,7 +745,18 @@ macro_rules! uint_impl {
         #[stable(feature = "rust1", since = "1.0.0")]
         #[inline]
         pub fn trailing_zeros(self) -> u32 {
-            unsafe { $cttz(self as $ActualT) as u32 }
+            // As of LLVM 3.6 the codegen for the zero-safe cttz8 intrinsic emits two conditional
+            // moves on x86_64. By promoting the value to u32 and setting bit 8, we get better code
```
Comment (alexcrichton) [inline src/libcore/num/mod.rs]: This comment may be a bit out of date (it's promoted to u16 now, right?). Also, can you wrap this to 80 chars as well? (surrounding style)

**Comment by Kimundi** [comment]
@alexcrichton:

Before:
```
test bench_cpp_naive_big     ... bench:  10324714 ns/iter (+/- 118672)
test bench_cpp_naive_short   ... bench:       139 ns/iter (+/- 22)
test bench_noop_big          ... bench:   1328772 ns/iter (+/- 38980)
test bench_noop_short        ... bench:        23 ns/iter (+/- 1)
test bench_rust_naive_big    ... bench:  15876210 ns/iter (+/- 618870)
test bench_rust_naive_short  ... bench:       167 ns/iter (+/- 5)
test bench_rust_unsafe_big   ... bench:  13832038 ns/iter (+/- 523291)
test bench_rust_unsafe_short ... bench:       139 ns/iter (+/- 13)
```
After:
```
test bench_cpp_naive_big     ... bench:  10378758 ns/iter (+/- 592938)
test bench_cpp_naive_short   ... bench:       124 ns/iter (+/- 5)
test bench_noop_big          ... bench:   1326851 ns/iter (+/- 19069)
test bench_noop_short        ... bench:        23 ns/iter (+/- 1)
test bench_rust_naive_big    ... bench:  15509013 ns/iter (+/- 877068)
test bench_rust_naive_short  ... bench:       168 ns/iter (+/- 5)
test bench_rust_unsafe_big   ... bench:  12895480 ns/iter (+/- 437780)
test bench_rust_unsafe_short ... bench:       128 ns/iter (+/- 0)
```

So there is a slight improvement, but its not as a significant as the numbers in the OP suggest.
(I also talked about it with @dotdash on IRC yesterday, my goal is for at least the unsafe code version to run at fast as the C++ one, and we haven't yet determined what teh cause for the difference is.)

**Comment by alexcrichton** [comment]
@Kimundi could you clarify on how you're running the benchmarks and the environment you're running in? I'm curious because I was able to reproduce the improvements.

**Comment by dotdash** [comment]
Link to patch added, comment updated and fitted to 80 chars wide (the rust code for vim were changed to a text width of 99, so auto-formatting messed that up).

**Comment by alexcrichton** [comment]
@bors: r+ 36dccec

Thanks! It looks like there may need to be some further investigation as @Kimundi's having trouble getting the same speedups, but this seems fine to go in for now at least!

**Comment by Kimundi** [comment]
My reproduction basically just doing `cargo clean; cargo bench` on a up-to-date arch linux system, x86_64, on an older toshiba laptop with i3 processor.

**Pull request merged**